### PR TITLE
HDDS-7661. Ratis Misreplication Handler

### PR DESCRIPTION
**Pull request by swamirishi**
## What changes were proposed in this pull request?

Add Ratis Misreplication Handler to solve misreplication for ratis based replicated containers.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7661

## How was this patch tested?
Unit Tests

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisMisReplicationHandler.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.hadoop.hdds.scm.node.NodeManager;
+import org.apache.hadoop.ozone.protocol.commands.ReplicateContainerCommand;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Handles the Ratis mis replication processing and forming the respective SCM
```
Comment (kerneltime) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisMisReplicationHandler.java:35]: I am looking at this for the first time; where is `mis replication` defined? The parent class also has a similar javadoc and https://issues.apache.org/jira/browse/HDDS-7661 is not linked to other relevant jiras.

Comment (swamirishi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisMisReplicationHandler.java:35]: Added javadoc

**Review by kerneltime — APPROVED** [review]

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestRatisMisReplicationHandler.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+  @BeforeEach
+  public void setup() {
+    RatisReplicationConfig repConfig = RatisReplicationConfig
+            .getInstance(ReplicationFactor.THREE);
+    setup(repConfig);
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {1, 2, 3, 4, 5, 6, 7})
```
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestRatisMisReplicationHandler.java:60]: Why more than 3 mis replication count?

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestRatisMisReplicationHandler.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  public void testMisReplicationWithAllNodesAvailable(int misreplicationCount)
+          throws IOException {
+    Set<ContainerReplica> availableReplicas = ReplicationTestUtil
+        .createReplicas(Pair.of(IN_SERVICE, 0), Pair.of(IN_SERVICE, 0),
+            Pair.of(IN_SERVICE, 0));
+    testMisReplication(availableReplicas, Collections.emptyList(),
+            0, misreplicationCount, Math.min(misreplicationCount, 3));
```
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestRatisMisReplicationHandler.java:67]: Just trying to understand: In which cases do we expect the number of target datanodes to be 3 for a ratis container?

**Review by siddhantsangwan — APPROVED** [review]
I just have some minor questions. Looks good overall.

**Pull request merged**